### PR TITLE
Db mk 1837 fabs valid congressional place performance

### PR DESCRIPTION
**Pull request by kimminahm**
https://federal-spending-transparency.atlassian.net/browse/DB-1837
AC: A and B
A) Non-USA FABS have a blank congressional District
B1) Derive CD from zip code
B2) Confirm that the district is a valid district

### CODE REVIEW

**Comment by alburde1** [comment]
[run tests]

File: dataactbroker/handlers/fileHandler.py [hunk 1]
```diff
@@ -1747,11 +1747,16 @@ def fabs_derivations(obj):
...
     if obj['funding_sub_tier_agency_co']:
         funding_sub_tier_agency_name = sess.query(SubTierAgency).\
             filter_by(sub_tier_agency_code=obj['funding_sub_tier_agency_co']).one()
         obj['funding_sub_tier_agency_na'] = funding_sub_tier_agency_name.sub_tier_agency_name
 
+    # deriving place of performance
+    if obj['place_of_performance_zip4a'] and not obj['place_of_performance_congr']:
+        zip_info = sess.query(Zips).\
+            filter_by(zip5=obj['place_of_performance_zip4a'][5:]).one()
+        obj['place_of_performance_congr'] = zip_info.congressional_district_no
```
Comment (alburde1) [inline dataactbroker/handlers/fileHandler.py]: What if we have a zip code in the format of `#########` or `#####-####`? There may be more than one congressional district that goes with a 5-digit zip (which is also why, if the format is `#####` you should do `first()` not `one()`). There's going to be some nested "if" statements here

File: dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_2.sql [hunk 2]
```diff
@@ -0,0 +1,37 @@
+--  If PrimaryPlaceOfPerformanceCode is not USA, Congressional District must be blank
```
Comment (alburde1) [inline dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_2.sql]: This comment doesn't match the content.

File: dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_2.sql [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    WHERE submission_id = {0})
+SELECT
+    dafa.row_number,
+    dafa.place_of_performance_zip4a,
+    dafa.place_of_performance_congr
+FROM detached_award_financial_assistance_d43_2_{0} AS dafa
+WHERE 
+	CASE WHEN COALESCE(dafa.place_of_performance_zip4a, '') != ''
+        -- 5 digit zip
+        THEN (CASE WHEN dafa.place_of_performance_zip4a ~ '^\d\d\d\d\d$'
```
Comment (alburde1) [inline dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_2.sql]: We aren't validating against the zip. The CD is only validated against the state

File: dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_2.sql [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    	place_of_performance_congr
+    FROM detached_award_financial_assistance
+    WHERE submission_id = {0})
+SELECT
+    dafa.row_number,
+    dafa.place_of_performance_zip4a,
+    dafa.place_of_performance_congr
+FROM detached_award_financial_assistance_d43_2_{0} AS dafa
+WHERE 
+	CASE WHEN COALESCE(dafa.place_of_performance_zip4a, '') != ''
```
Comment (alburde1) [inline dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_2.sql]: This part doesn't need to be a CASE. We only need to check here that it's provided.

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def test_success(database):
+    """ If PrimaryPlaceOfPerformanceCode is not USA, Congressional District must be blank """
+
+    det_award_1 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="Nk",
+                                                          place_of_performance_congr="")
+    det_award_2 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="CA",
+                                                          place_of_performance_congr="")
+    det_award_3 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="Mx",
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py]: Change one of these from `""` to `None` to make sure it works

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                                                          place_of_performance_congr="")
+    det_award_4 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="USA",
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py]: Check this with a different capitalization of this as well. `usa` or `UsA` or something, just not all caps

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                                                          place_of_performance_congr="12")
+    errors = number_of_errors(_FILE, database, models=[det_award_1, det_award_2, det_award_3, det_award_4])
+    assert errors == 0
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py]: Add tests to check for country code of `usa` and CD of `""` and `None`

File: dataactbroker/handlers/fileHandler.py (hunk 1 above)
Comment (kimminahm) [inline dataactbroker/handlers/fileHandler.py]: regardless of whether they use a dash or not, the [5:] will grab the first 5 digits of the zipcode which should be before any other symbol. I will seitch from one to first

File: dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_1.sql [hunk 8]
```diff
@@ -0,0 +1,10 @@
+--  If PrimaryPlaceOfPerformanceCode is not USA, Congressional District must be blank
+
+SELECT
+    row_number,
+    place_of_performance_code,
+    place_of_perform_country_c
+FROM detached_award_financial_assistance
+WHERE submission_id = {0}
+	AND UPPER(place_of_perform_country_c) != 'USA'
```
Comment (alburde1) [inline dataactvalidator/config/sqlrules/d43_detached_award_financial_assistance_1.sql]: Tabbing

File: dataactvalidator/config/sqlrules/sqlRules.csv [hunk 9]
```diff
@@ -156,4 +156,6 @@ D41,"For PrimaryPlaceOfPerformanceCode XX##### city must exist in provided state
...
 D42,"PrimaryPlaceOfPerformanceForeignLocationDescription is required for foreign places of performance (i.e., when PrimaryPlaceOfPerformanceCountryCode does not equal USA).","PrimaryPlaceOfPerformanceForeignLocationDescription is required for foreign places of performance (i.e., when PrimaryPlaceOfPerformanceCountryCode does not equal USA).",n,detached_award,fatal,,d42_detached_award_financial_assistance_1
 D42,"PrimaryPlaceOfPerformanceForeignLocationDescription must be blank for domestic recipients (i.e., when PrimaryPlaceOfPerformanceCountryCode = USA).","PrimaryPlaceOfPerformanceForeignLocationDescription must be blank for domestic recipients (i.e., when PrimaryPlaceOfPerformanceCountryCode = USA).",n,detached_award,fatal,,d42_detached_award_financial_assistance_2
+D43,"When primaryplaceofperformancecountrycode is not USA Congressional District must be empty","primaryplaceofperformancecongressionaldistrict must be blank when primaryplaceofperformancecountrycode is not USA",n,detached_award,fatal,,d43_detached_award_financial_assistance_1
```
Comment (alburde1) [inline dataactvalidator/config/sqlrules/sqlRules.csv]: Capitalize primaryplaceofperformancecountrycode and primaryplaceofperformancecongressionaldistrict  as they are in the ticket and in other places

File: dataactvalidator/config/sqlrules/sqlRules.csv [hunk 10]
```diff
@@ -156,4 +156,6 @@ D41,"For PrimaryPlaceOfPerformanceCode XX##### city must exist in provided state
...
 D42,"PrimaryPlaceOfPerformanceForeignLocationDescription must be blank for domestic recipients (i.e., when PrimaryPlaceOfPerformanceCountryCode = USA).","PrimaryPlaceOfPerformanceForeignLocationDescription must be blank for domestic recipients (i.e., when PrimaryPlaceOfPerformanceCountryCode = USA).",n,detached_award,fatal,,d42_detached_award_financial_assistance_2
+D43,"When primaryplaceofperformancecountrycode is not USA Congressional District must be empty","primaryplaceofperformancecongressionaldistrict must be blank when primaryplaceofperformancecountrycode is not USA",n,detached_award,fatal,,d43_detached_award_financial_assistance_1
+D43,"PrimaryPlaceOfPerformanceCongressionalDistrict must be provided and valid when PrimaryPlaceofPerformanceZIP+4 is provided","PrimaryPlaceOfPerformanceCongressionalDistrict must be provided and valid when PrimaryPlaceofPerformanceZIP+4 is provided",n,detached_award,fatal,,d43_detached_award_financial_assistance_2
```
Comment (alburde1) [inline dataactvalidator/config/sqlrules/sqlRules.csv]: This is not accurate, PrimaryPlaceOfPerformanceCongressionalDistrict is NOT required when PrimaryPlaceofPerformanceZIP+4 is provided.

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+def test_success(database):
+    """ If PrimaryPlaceOfPerformanceCode is not USA, Congressional District must be blank """
+
+    det_award_1 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="Nk",
+                                                          place_of_performance_congr="")
+    det_award_2 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="CA",
+                                                          place_of_performance_congr=None)
+    det_award_3 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="Usa",
+                                                          place_of_performance_congr="")
+    det_award_4 = DetachedAwardFinancialAssistanceFactory(place_of_perform_country_c="USA",
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py]: Since these two USA entries are using different CD contents (one is there and one isn't) include a USA entry not in all caps that has a CD

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+                                                          place_of_performance_congr="12")
+    errors = number_of_errors(_FILE, database, models=[det_award_1, det_award_2, det_award_3, det_award_4])
+    assert errors == 0
+
+
+def test_failure(database):
+    """ Test failure for PrimaryPlaceOfPerformanceCode for aggregate records (i.e., when RecordType = 1)
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_1.py]: This is not accurate

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_2.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+def test_column_headers(database):
+    expected_subset = {"row_number", "place_of_performance_zip4a", "place_of_performance_congr",
+                       "place_of_performance_code"}
+    actual = set(query_columns(_FILE, database))
+    print(actual)
+    assert expected_subset == actual
+
+
+def test_success(database):
+    """ When provided, PrimaryPlaceofPerformanceZIP+4, placeOfPerformanceCongressCode must be in the state specified
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_2.py]: This sentence is a little unclear

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_2.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    # valid 9 digit zip
+    det_award_5 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="123456789",
+                                                          place_of_performance_congr="04",
+                                                          place_of_performance_code="NY*****")
+
+    det_award_6 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="123456780",
+                                                          place_of_performance_congr="03",
+                                                          place_of_performance_code="NY*****")
+
+    det_award_7 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="12345-6781",
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_2.py]: You don't have to check for all types of zips, that's not what this rule is testing at all. What we need to look at is different district numbers for different states (upper and lowercase) and make sure it's validating all the cases. Same comment for failures

File: tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_2.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+                                                          place_of_performance_congr="02",
+                                                          place_of_performance_code="NY*****")
+
+    errors = number_of_errors(_FILE, database, models=[det_award_1, det_award_2, det_award_3, det_award_4, det_award_5,
+                                                       det_award_6, det_award_7, zips1, zips2, zips3, zips4])
+    assert errors == 0
+
+
+def test_failure(database):
+    """ Test failure for PrimaryPlaceOfPerformanceCongressCode not in PrimaryPlaceOfPerformanceCode or not provided
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_2.py]: It's allowed to not be provided, that's not part of the validation
